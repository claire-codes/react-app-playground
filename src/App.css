.App {
  text-align: center;
}

.App-logo {
  animation: App-logo-spin infinite 20s linear;
  height: 80px;
}

.App-header {
  background-color: #222;
  height: 150px;
  padding: 20px;
  color: white;
}

.App-intro {
  font-size: large;
}

@keyframes App-logo-spin {
  from { transform: rotate(0deg); }
  to { transform: rotate(360deg); }
}

input[type=number]::-webkit-inner-spin-button, 
input[type=number]::-webkit-outer-spin-button { 
  -webkit-appearance: none; 
  margin: 0; 
}
