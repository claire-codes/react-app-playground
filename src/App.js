import React, { Component } from "react";
import styled from "styled-components";

import Foo from "./Foo";
import Widget from "./Widget";

import weightConverter from "./weightConverter";
import { ReactComponent as Logo } from "./logo.svg";
import "./App.css";

const InputHolder = styled.div`
  margin: 20px;
`;

const InputField = styled.input`
  background-color: #61dafb;
  border: none;
  border-bottom: 1px solid black;
  font-size: 20px;
  text-align: center;
`;

const InputWrapper = styled.div`
  display: inline-block;
  margin-bottom: 15px;
  margin-left: 10px;
`;

class App extends Component {
  constructor(props) {
    super(props);
    this.state = {
      kg: 0,
      stone: 0,
      lbs: 0
    };

    this.handleKgChange = this.handleKgChange.bind(this);
    this.handleStoneChange = this.handleStoneChange.bind(this);
    this.handleLbsChange = this.handleLbsChange.bind(this);
  }

  handleKgChange(event) {
    this.setState({ kg: event.target.value });
    if (event.target.value) {
      let tmp = weightConverter.kgToStoneLbs(event.target.value);
      this.setState({ stone: tmp.stone, lbs: tmp.lbs });
    }
  }

  handleStoneChange(event) {
    this.setState({ stone: event.target.value });
    if (event.target.value) {
      let tmp = weightConverter.stoneLbsToKg(
        event.target.value,
        this.state.lbs
      );
      this.setState({ kg: tmp });
    }
  }

  handleLbsChange(event) {
    this.setState({ lbs: event.target.value });
    if (event.target.value) {
      let tmp = weightConverter.stoneLbsToKg(
        this.state.stone,
        event.target.value
      );
      this.setState({ kg: tmp });
    }
  }

  componentDidMount() {
    if (this.nameInput) this.nameInput.focus();
  }

  render() {
    return (
      <div className="App">
        <div className="App-header">
          <Logo className="App-logo" />
          <h2 className="qa-header">Convert kg to stone</h2>
        </div>
        <InputHolder>
          <InputHolder>
            <InputField
              type="number"
              onChange={this.handleKgChange}
              value={this.state.kg}
              ref={input => {
                this.nameInput = input;
              }}
            />
            &nbsp;kg
          </InputHolder>
          <InputWrapper>
            <InputField
              type="number"
              onChange={this.handleStoneChange}
              value={this.state.stone}
            />
            &nbsp;stone
          </InputWrapper>
          <InputWrapper>
            <InputField
              type="number"
              onChange={this.handleLbsChange}
              value={this.state.lbs}
              className="qa-lbs-input"
            />
            &nbsp;lbs
          </InputWrapper>
        </InputHolder>
        <Foo invisible={true}>
          <h2>Hello World</h2>
        </Foo>
        <Foo>
          <h2>Goodbye Cruel World</h2>
        </Foo>
        <Widget invisible>Fancy pants child prop</Widget>
        <Widget>Text is a child prop</Widget>
      </div>
    );
  }
}

export default App;
