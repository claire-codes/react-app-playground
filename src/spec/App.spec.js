import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import Enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
Enzyme.configure({ adapter: new Adapter() });
import 'jest-styled-components'

import { mount, shallow } from 'enzyme';
import App from '../App';

it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<App />, div);
});

describe('snapshot tests', () => {
    it('Snapshot on <App /> | react-test-renderer', () => {
        const component = renderer.create(<App />).toJSON();
        expect(component).toMatchSnapshot();
    });

    it('Snapshot on <App /> | enzyme shallow', () => {
        const component = shallow(<App />);
        expect(component).toMatchSnapshot();
    });

    // I tried a mount but it seemed to make the tests hang
});

describe('Enzyme tests', () => {
    // These first two are testing that it exists in the structure we expect
    it('renders header text', () => {
        const app = shallow(<App />);
        expect(app.find('.qa-header').text()).toEqual('Convert kg to stone');
    });

    it('has exactly three input fields', () => {
        const app = mount(<App />);
        expect(app.find('input[type="number"]').length).toEqual(3);
    });
});

describe('Behaviour tests', () => {
    let app;

    beforeEach(() => {
        app = mount(<App />);
    });

    it('renders text and state is OK', () => {
        expect(app.find('.App-header').first()).toBeDefined();
        expect(app.state('kg')).toEqual(0);
        expect(app.state('stone')).toEqual(0);
        expect(app.state('lbs')).toEqual(0);
    });

    it('calls handleLbsChange when value changes', () => {
        const lbsInput = app.find('.qa-lbs-input').first();
        lbsInput.simulate('change', { target: { value: 2 } });
        expect(app.state('kg')).toEqual(0.89);
        expect(app.state('stone')).toEqual(0);
        expect(app.state('lbs')).toEqual(2);
    });
});
