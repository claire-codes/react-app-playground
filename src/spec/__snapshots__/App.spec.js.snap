// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot tests Snapshot on <App /> | enzyme shallow 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <App />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="App-header"
>
          <img
                    alt="logo"
                    className="App-logo"
                    src="logo.svg"
          />
          <h2>
                    Convert kg to stone
          </h2>
</div>,
        <styled.div>
          <styled.div>
                    <styled.input
                              innerRef={[Function]}
                              onChange={[Function]}
                              type="number"
                              value={0}
                    />
                     kg
          </styled.div>
          <styled.div>
                    <styled.input
                              onChange={[Function]}
                              type="number"
                              value={0}
                    />
                     stone
          </styled.div>
          <styled.div>
                    <styled.input
                              onChange={[Function]}
                              type="number"
                              value={0}
                    />
                     lbs
          </styled.div>
</styled.div>,
        <styled.div
          invisible={true}
>
          <h2>
                    Hello World
          </h2>
</styled.div>,
        <styled.div>
          <h2>
                    Goodbye Cruel World
          </h2>
</styled.div>,
        <Widget
          invisible={true}
>
          Fancy pants child prop
</Widget>,
        <Widget>
          Text is a child prop
</Widget>,
      ],
      "className": "App",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <img
              alt="logo"
              className="App-logo"
              src="logo.svg"
/>,
            <h2>
              Convert kg to stone
</h2>,
          ],
          "className": "App-header",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "alt": "logo",
              "className": "App-logo",
              "src": "logo.svg",
            },
            "ref": null,
            "rendered": null,
            "type": "img",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Convert kg to stone",
            },
            "ref": null,
            "rendered": "Convert kg to stone",
            "type": "h2",
          },
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": Array [
            <styled.div>
              <styled.input
                            innerRef={[Function]}
                            onChange={[Function]}
                            type="number"
                            value={0}
              />
               kg
</styled.div>,
            <styled.div>
              <styled.input
                            onChange={[Function]}
                            type="number"
                            value={0}
              />
               stone
</styled.div>,
            <styled.div>
              <styled.input
                            onChange={[Function]}
                            type="number"
                            value={0}
              />
               lbs
</styled.div>,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": Array [
                <styled.input
                  innerRef={[Function]}
                  onChange={[Function]}
                  type="number"
                  value={0}
/>,
                " kg",
              ],
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "innerRef": [Function],
                  "onChange": [Function],
                  "type": "number",
                  "value": 0,
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              " kg",
            ],
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": Array [
                <styled.input
                  onChange={[Function]}
                  type="number"
                  value={0}
/>,
                " stone",
              ],
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "onChange": [Function],
                  "type": "number",
                  "value": 0,
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              " stone",
            ],
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": Array [
                <styled.input
                  onChange={[Function]}
                  type="number"
                  value={0}
/>,
                " lbs",
              ],
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "onChange": [Function],
                  "type": "number",
                  "value": 0,
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              " lbs",
            ],
            "type": [Function],
          },
        ],
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": <h2>
            Hello World
</h2>,
          "invisible": true,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Hello World",
          },
          "ref": null,
          "rendered": "Hello World",
          "type": "h2",
        },
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": <h2>
            Goodbye Cruel World
</h2>,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Goodbye Cruel World",
          },
          "ref": null,
          "rendered": "Goodbye Cruel World",
          "type": "h2",
        },
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": "Fancy pants child prop",
          "invisible": true,
        },
        "ref": null,
        "rendered": "Fancy pants child prop",
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": "Text is a child prop",
        },
        "ref": null,
        "rendered": "Text is a child prop",
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="App-header"
>
            <img
                        alt="logo"
                        className="App-logo"
                        src="logo.svg"
            />
            <h2>
                        Convert kg to stone
            </h2>
</div>,
          <styled.div>
            <styled.div>
                        <styled.input
                                    innerRef={[Function]}
                                    onChange={[Function]}
                                    type="number"
                                    value={0}
                        />
                         kg
            </styled.div>
            <styled.div>
                        <styled.input
                                    onChange={[Function]}
                                    type="number"
                                    value={0}
                        />
                         stone
            </styled.div>
            <styled.div>
                        <styled.input
                                    onChange={[Function]}
                                    type="number"
                                    value={0}
                        />
                         lbs
            </styled.div>
</styled.div>,
          <styled.div
            invisible={true}
>
            <h2>
                        Hello World
            </h2>
</styled.div>,
          <styled.div>
            <h2>
                        Goodbye Cruel World
            </h2>
</styled.div>,
          <Widget
            invisible={true}
>
            Fancy pants child prop
</Widget>,
          <Widget>
            Text is a child prop
</Widget>,
        ],
        "className": "App",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <img
                alt="logo"
                className="App-logo"
                src="logo.svg"
/>,
              <h2>
                Convert kg to stone
</h2>,
            ],
            "className": "App-header",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "alt": "logo",
                "className": "App-logo",
                "src": "logo.svg",
              },
              "ref": null,
              "rendered": null,
              "type": "img",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Convert kg to stone",
              },
              "ref": null,
              "rendered": "Convert kg to stone",
              "type": "h2",
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": Array [
              <styled.div>
                <styled.input
                                innerRef={[Function]}
                                onChange={[Function]}
                                type="number"
                                value={0}
                />
                 kg
</styled.div>,
              <styled.div>
                <styled.input
                                onChange={[Function]}
                                type="number"
                                value={0}
                />
                 stone
</styled.div>,
              <styled.div>
                <styled.input
                                onChange={[Function]}
                                type="number"
                                value={0}
                />
                 lbs
</styled.div>,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": Array [
                  <styled.input
                    innerRef={[Function]}
                    onChange={[Function]}
                    type="number"
                    value={0}
/>,
                  " kg",
                ],
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "class",
                  "props": Object {
                    "innerRef": [Function],
                    "onChange": [Function],
                    "type": "number",
                    "value": 0,
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
                " kg",
              ],
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": Array [
                  <styled.input
                    onChange={[Function]}
                    type="number"
                    value={0}
/>,
                  " stone",
                ],
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "class",
                  "props": Object {
                    "onChange": [Function],
                    "type": "number",
                    "value": 0,
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
                " stone",
              ],
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": Array [
                  <styled.input
                    onChange={[Function]}
                    type="number"
                    value={0}
/>,
                  " lbs",
                ],
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "class",
                  "props": Object {
                    "onChange": [Function],
                    "type": "number",
                    "value": 0,
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
                " lbs",
              ],
              "type": [Function],
            },
          ],
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": <h2>
              Hello World
</h2>,
            "invisible": true,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Hello World",
            },
            "ref": null,
            "rendered": "Hello World",
            "type": "h2",
          },
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": <h2>
              Goodbye Cruel World
</h2>,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Goodbye Cruel World",
            },
            "ref": null,
            "rendered": "Goodbye Cruel World",
            "type": "h2",
          },
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": "Fancy pants child prop",
            "invisible": true,
          },
          "ref": null,
          "rendered": "Fancy pants child prop",
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": "Text is a child prop",
          },
          "ref": null,
          "rendered": "Text is a child prop",
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;

exports[`snapshot tests Snapshot on <App /> | react-test-renderer 1`] = `
<div
  className="App"
>
  <div
    className="App-header"
  >
    <img
      alt="logo"
      className="App-logo"
      src="logo.svg"
    />
    <h2>
      Convert kg to stone
    </h2>
  </div>
  <div
    className="sc-htpNat eusUoO"
  >
    <div
      className="sc-htpNat eusUoO"
    >
      <input
        className="sc-bxivhb ihvhmd"
        onChange={[Function]}
        type="number"
        value={0}
      />
       kg
    </div>
    <div
      className="sc-ifAKCX ftoKPH"
    >
      <input
        className="sc-bxivhb ihvhmd"
        onChange={[Function]}
        type="number"
        value={0}
      />
       stone
    </div>
    <div
      className="sc-ifAKCX ftoKPH"
    >
      <input
        className="sc-bxivhb ihvhmd"
        onChange={[Function]}
        type="number"
        value={0}
      />
       lbs
    </div>
  </div>
  <div
    className="sc-bdVaJa empOPp"
  >
    <h2>
      Hello World
    </h2>
  </div>
  <div
    className="sc-bdVaJa ivQMso"
  >
    <h2>
      Goodbye Cruel World
    </h2>
  </div>
  <div
    className="sc-bwzfXH ZZYNN"
  >
    Fancy pants child prop
  </div>
  <div
    className="sc-bwzfXH fTfPYm"
  >
    Text is a child prop
  </div>
</div>
`;
